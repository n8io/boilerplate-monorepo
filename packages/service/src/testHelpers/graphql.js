import { PubSub } from 'apollo-server-express';
import { createTestClient } from 'apollo-server-testing';
import { subscribe as graphqlSubscribe } from 'graphql';
import { has, head, keys, path, pipe, prop, __ } from 'ramda';
import { makeServer } from '../server';
import { make as makeContext } from './context';

const makeGraphqlClient = (context) => {
  const server = makeServer({ context });
  const { query, mutate } = createTestClient(server);

  const subscribe = async (subscription, pubsub) => {
    const iterator = await graphqlSubscribe(
      server.schema,
      subscription,
      {},
      { ...context, pubsub }
    );

    if (has('errors', iterator)) return iterator;

    const { value } = await iterator.next();

    return value;
  };

  const whileSubscribed = (subscription, callback) => {
    const pubsub = new PubSub();
    const promise = subscribe(subscription, pubsub);

    callback(pubsub);
    return promise;
  };

  return {
    context,
    execMutation: mutate,
    execQuery: query,
    whileSubscribed,
  };
};

const graphqlClient = makeGraphqlClient(makeContext());

const responseToData = (response) => {
  // eslint-disable-next-line no-console
  response.errors && console.error(response.errors);
  expect(response.errors).toBeUndefined();
  expect(response.data).not.toBeNull();

  return pipe(keys, head, prop(__, response.data))(response.data);
};

const responseToErrorCode = path(['errors', 0, 'extensions', 'code']);

export {
  graphqlClient,
  makeGraphqlClient,
  responseToData,
  responseToErrorCode,
};
