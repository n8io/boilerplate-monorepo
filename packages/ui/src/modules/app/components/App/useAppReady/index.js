import { config } from 'config';
import { useEffect, useState } from 'react';
import { AccessToken } from 'types/accessToken';
import { useIsColdStarting } from '../useIsColdStarting';

const useAppReady = () => {
  const isColdStarting = useIsColdStarting();
  const [isReady, setReady] = useState(false);

  useEffect(() => {
    if (isColdStarting) return;

    const uri = new URL(config.GRAPHQL_URI);

    uri.pathname = 'refresh_token';

    fetch(uri.href, {
      credentials: 'include',
      method: 'POST',
    })
      .then(async (payload) => {
        const data = await payload.json();
        const { ok, token } = data;

        ok && AccessToken.set(token);
      })
      .finally(() => setReady(true));
  }, [isColdStarting]);

  return !isColdStarting && isReady;
};

export { useAppReady };
