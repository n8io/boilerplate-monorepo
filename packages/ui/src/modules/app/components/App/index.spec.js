import React from 'react';
import { render } from 'testHelpers';
import { App } from '.';
import * as UseAppReadyHook from './useAppReady';

jest.mock('./Overlays/ColdStart', () => ({
  ColdStart: (props) => <x-ColdStart {...props} data-testid="cold-start" />,
}));

jest.mock('./Layout', () => ({
  Layout: (props) => <x-Layout {...props} data-testid="layout" />,
}));

jest.mock('./Providers', () => ({
  Providers: (props) => <x-Providers {...props} data-testid="providers" />,
}));

describe('<App/>', () => {
  const defaultProps = {};

  const renderComponent = (overrides) =>
    render(<App {...defaultProps} {...overrides} />);

  describe('when app is ready', () => {
    const isAppReady = true;

    beforeEach(() => {
      const useAppReady = td.replace(UseAppReadyHook, 'useAppReady');

      td.when(useAppReady()).thenReturn(isAppReady);
    });

    test('renders its children', () => {
      const { container, getByTestId } = renderComponent();

      expect(getByTestId('providers')).not.toBeNull();
      expect(getByTestId('providers')).not.toBeNull();
      expect(container.firstChild).toMatchSnapshot();
    });
  });

  describe('when app is NOT ready', () => {
    const isAppReady = false;

    beforeEach(() => {
      const useAppReady = td.replace(UseAppReadyHook, 'useAppReady');

      td.when(useAppReady()).thenReturn(isAppReady);
    });

    test('does NOT renders its children', () => {
      const { queryByTestId } = renderComponent();

      expect(queryByTestId('providers')).toBeNull();
      expect(queryByTestId('providers')).toBeNull();
    });
  });
});
