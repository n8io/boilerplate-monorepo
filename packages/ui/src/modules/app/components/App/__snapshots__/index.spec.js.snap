// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App/> when app is ready renders its children 1`] = `
<x-coldstart
  data-testid="cold-start"
/>
`;
